packages/permission-log-controller/src/PermissionLog.ts

```diff
@@ -0,0 +1,504 @@
+        : LOG_METHOD_TYPES.restricted,
+      origin: request.origin,
+      requestTime: Date.now(),
+      responseTime: null,
+      success: null,
+    };
+    this.commitNewActivity(activityEntry);
+    return activityEntry;
+  }
+
+  /**
+   * Adds response data to an existing activity log entry.
+   * Entry assumed already committed (i.e., in the log).
+   *
+   * @param entry - The entry to add a response to.
+   * @param response - The response object.
+   * @param time - Output from Date.now()
+   */
+  logResponse(
+    entry: PermissionActivityLog,
+    response: PendingJsonRpcResponse<(string | Permission)[]>,
+    time: number,
+  ) {
+    if (!entry || !response) {
+      return;
+    }
+
+    // The JSON-RPC 2.0 specification defines "success" by the presence of
+    // either the "result" or "error" property. The specification forbids
+    // both properties from being present simultaneously, and our JSON-RPC
+    // stack is spec-compliant at the time of writing.
+    this.update((state) => {
+      state.permissionActivityLog[state.permissionActivityLog.length - 1] = {
+        ...entry,
+        success: Object.hasOwnProperty.call(response, 'result'),
+        responseTime: time,
+      };
+    });
+  }
+
+  /**
+   * Commit a new entry to the activity log.
+   * Removes the oldest entry from the log if it exceeds the log limit.
+   *
+   * @param entry - The activity log entry.
+   */
+  commitNewActivity(entry: PermissionActivityLog) {
+    const logs = [...this.getActivityLog(), entry];
+
+    // remove oldest log if exceeding size limit
+    if (logs.length > LOG_LIMIT) {
+      logs.shift();
+    }
+
+    this.updateActivityLog(logs);
+  }
+
+  /**
+   * Create new permissions history log entries, if any, and commit them.
+   *
+   * @param requestedMethods - The method names corresponding to the requested permissions.
+   * @param origin - The origin of the permissions request.
+   * @param result - The permissions request response.result.
+   * @param time - The time of the request, i.e. Date.now().
+   * @param isEthRequestAccounts - Whether the permissions request was 'eth_requestAccounts'.
+   */
+  logPermissionsHistory(
+    requestedMethods: string[],
+    origin: string,
+    result: (string | Permission)[],
+    time: number,
+    isEthRequestAccounts: boolean,
+  ) {
+    let accounts: string[];
+    let newEntries: PermissionEntry;
+
+    if (isEthRequestAccounts) {
+      accounts = result as string[];
+      const accountToTimeMap = getAccountToTimeMap(accounts, time);
+
+      newEntries = {
+        eth_accounts: {
+          accounts: accountToTimeMap,
+          lastApproved: time,
+        },
+      };
+    } else {
+      // Records new "lastApproved" times for the granted permissions, if any.
+      // Special handling for eth_accounts, in order to record the time the
+      // accounts were last seen or approved by the origin.
+      newEntries = (result as Permission[])
```

Comment:
Would you mind adding a comment explaining why this type assertion is necessary?

``` suggestion
      // Type assertion: We are assuming the type of the response data if the
      // RPC method is "wallet_*" or within the set of restricted RPC methods.
      newEntries = (result as Permission[])